gems/smithy/lib/smithy/vise/operation_index.rb

```diff
@@ -2,73 +2,18 @@
 
 module Smithy
   module Vise
-    # Finds all operation shapes in a service.
+    # Finds all shapes for an operation.
     class OperationIndex
-      RESOURCE_LIFECYCLE_KEYS = %w[create put read update delete list].freeze
-
       # @param [Hash] model Model
       def initialize(model)
-        @shapes = model['shapes']
-      end
-
-      # @param [Hash] service Service shape
-      # @return [Hash<String, Hash>] The operations for the service.
-      def for(service)
-        operations = {}
-        _id, service = service.first
-        parse_service_operations(service, operations)
-        parse_service_resources(service, operations)
-        operations.sort_by { |k, _v| k }.to_h
-      end
-
-      private
-
-      def parse_service_operations(service, operations)
-        service['operations']&.collect do |shape|
-          id = shape['target']
-          operations[id] = @shapes[id]
-        end
-      end
-
-      def parse_service_resources(service, operations)
-        service['resources']&.collect do |shape|
-          id = shape['target']
-          parse_resource(@shapes[id], operations)
-        end
-      end
-
-      def parse_resource(resource, operations)
-        parse_lifecycles(resource, operations)
-        parse_resource_operations(resource, operations)
-        parse_resource_collection_operations(resource, operations)
-
-        resource['resources']&.collect do |shape|
-          id = shape['target']
-          parse_resource(@shapes[id], operations)
-        end
-      end
-
-      def parse_lifecycles(resource, operations)
-        RESOURCE_LIFECYCLE_KEYS.each do |key|
-          next unless resource[key]
-
-          id = resource[key]['target']
-          operations[id] = @shapes[id]
-        end
-      end
-
-      def parse_resource_operations(resource, operations)
-        resource['operations']&.collect do |shape|
-          id = shape['target']
-          operations[id] = @shapes[id]
-        end
+        @operation_parser = OperationParser.new(model)
+        @memo = {}
       end
 
-      def parse_resource_collection_operations(resource, operations)
-        resource['collectionOperations']&.collect do |shape|
-          id = shape['target']
-          operations[id] = @shapes[id]
-        end
+      # @param [Hash] operation Operation shape
+      # @return [Hash<String, Hash>] The shapes for the operation.
+      def shapes_for(operation)
+        @memo[operation] ||= @operation_parser.shapes_for(operation)
```

Comment:
Is this class just a cache on top of the operation parser? It feels like it doesn't really need to exist and this memoization/caching could just be done in the OperationParser.

Reply:
Yeah, this one isn't quite useful. I had a thought last night to actually rewrite this index to get input and output shapes, and potentially a shape walker that does a depth first traversal of an operation. That would be useful for generating request and response documentation.